Fixes #100: Add group instantiation advice (#121)
cfrg/draft-irtf-cfrg-voprf: draft-irtf-cfrg-voprf.md

## Non-Verifiable Ciphersuites

### OPRF-curve25519\_XMD:SHA-512\_ELL2_RO\_: {#oprf-c25519}

- Group instantiation:
  - Elliptic curve: curve25519 {{RFC7748}}

>   - Generator(): Return the point with the following hex coordinates:

Reviewer: Should we clarify that these are affine coordinates (and perhaps cite RFC7748, too)?

    - x = `09`
    - y = `20AE19A1B8A086B4E01EDD2C7748D14C923D4D7E6D7C61B229E9C5A27ECED3D9`
  - HashToGroup(): curve25519_XMD:SHA-512_ELL2_RO_ {{I-D.irtf-cfrg-hash-to-curve}} with DST "RFCXXXX-VOPRF-curve25519_XMD:SHA-512_ELL2_RO_"
  - Serialization: The standard 32-byte representation of the public key {{!RFC7748}}

>   - Order(): Returns `2^225 - 19`

Reviewer 2: This number should be in hex too.

  - Addition: Adding curve points directly corresponds to the group
    addition operation.
  - Scalar multiplication: Implementers must validate all untrusted

>     input points to the scalar multiplication algorithm. Validation

Reviewer 2: Validation is not as easy as all-zero checking. I think it's better to add that "points must be in the prime order subgroup".

    and aborting the protocol if such an output is found. See
    {{RFC7748}} for more details.
- H1: SHA512

### OPRF-curve448\_XMD:SHA-512\_ELL2\_RO\_: {#oprf-c448}

- Group instantiation:
  - Elliptic curve: curve448 {{RFC7748}}
  - Generator(): Return the point with the following affine coordinates:
    - x = `05`
    - y = `7D235D1295F5B1F66C98AB6E58326FCECBAE5D34F55545D060F75DC28DF3F6EDB8027E2346430D211312C4B150677AF76FD7223D457B5B1A`
  - HashToGroup(): curve448_XMD:SHA-512_ELL2_RO_ {{I-D.irtf-cfrg-hash-to-curve}} with DST "RFCXXXX-OPRF-curve448_XMD:SHA-512_ELL2_RO_"
  - Serialization: The standard 56-byte representation of the public key {{!RFC7748}}
  - Order(): Returns `2^448 - 2^224 - 1`
  - Addition: Adding curve points directly corresponds to the group
    addition operation.
  - Scalar multiplication: Implementers must validate all untrusted
    input points to the scalar multiplication algorithm. Validation
    consists of checking for all-zero outputs of scalar multiplication
    and aborting the protocol if such an output is found. See
    {{RFC7748}} for more details.
- H1: SHA512

### OPRF-P256\_XMD:SHA-256\_SSWU\_RO\_: {#oprf-p256}

- Group instantiation:

>   - Elliptic curve: secp256r1 {{SEC2}}

Reviewer 2: ANSI X9.62 is the usual reference for elliptic curves.

- H1: SHA512
- H2: HKDF-Expand-SHA512
- H3: SHA512

## Group instantiations

We use ciphersuites based on elliptic curves to instantiate the protocol
from {{protocol}}. We discuss exactly how the definition of these curves
corresponds to the prime-order group API in {{pog}}.

### NIST curves

For the NIST curves: P256, P384, and P521; the instantiation of a
prime-order group is relatively straightforward. In particular, the
cofactor of the curve is 1, which means that no special attention to
small-subgroup attacks has to be taken into account.

Roughly speaking, any valid point on the curve corresponds directly to a
group element. Adding curve points corresponds to point addition, and
scalar multiplication algorithms for curve points corresponds to scalar
multiplication in the group. The `HashToGroup`, `Serialize` and
`Deserialize` algorithms are handled as explained in the ciphersuite
definition. The generators of each of the groups are given in {{SEC2}}.
The identity element of the group is the point at infinity. Finally, the
`Order` algorithm returns the order of points in the group.

### curve25519 and curve448

> These curves come with extra considerations surrounding scalar
> multiplication since the cofactor of the curve is now `4`. This means

Reviewer: ```suggestion
curve25519 and curve448 come with extra considerations surrounding scalar
multiplication since the cofactor of each is `4`. This means
```

Reviewer 2: since the cofactor is `8` and `4`, respectively.

> that curve points can either belong to a small subgroup of order `4`, or
> a larger prime-order subgroup. The larger prime-order subgroup

Reviewer: ```suggestion
that curve points can either belong to a small subgroups that are not of 
prime order. The larger prime-order subgroup
```

essentially corresponds to the prime-order group that the (V)OPRF takes

> place in. To ensure that all operations are performed in this subgroup,

Reviewer 2: What is the reasoning supporting this paragraph?

Reviewer: Can you elaborate on the question?

Reviewer 2: >To ensure that all operations are performed in this subgroup, any implementer must check for all-zero outputs of the scalar multiplication algorithms wherever they are used.

I don't see the relation between performing operations in the subgroup versus cheking for all-zeros.

*Recall that the procedures described in rfc7748 are specially crafted for a Diffie-Hellman protocol.
